stock_request/models/stock_request_order.py

```diff
@@ -277,14 +277,17 @@ def action_view_stock_requests(self):
             action["res_id"] = self.stock_request_ids.id
         return action
 
-    @api.model
-    def create(self, vals):
-        upd_vals = vals.copy()
-        if upd_vals.get("name", "/") == "/":
-            upd_vals["name"] = self.env["ir.sequence"].next_by_code(
-                "stock.request.order"
-            )
-        return super().create(upd_vals)
+    @api.model_create_multi
+    def create(self, vals_list):
+        vals_list_upd = []
+        for vals in vals_list:
+            upd_vals = vals.copy()
```

Comment:
No need of copy AFAIK

Reply:
In fact, this is better as the variable argument is a pointer to values. Changing them directly here can cause problems in the upper stack.

Reply:
I don't see any problem here in the "upper stack". Can you point to any real example of a problem? This is causing a non-needed overhead IMO.

Reply:
Although I have changed what you suggest, I **BELIEVE** I know what the problem is.

**Example 1**:
```
vals = {"product_id": 1}
request_1 = self.env["stock.request"].create(vals)
request_2 = self.env["stock.request"].create(vals)
```

We will get an error because when trying to create the request_2 vals will already have set `name` according to the sequence, it will try to create the request with the same name and that is not allowed.


**Example 2 (similar to the previous one)**:

```
vals = ({"product_id": 1},{"product_id": 1}}
requests = self.env["stock.request"].create(vals)
```

We will get an error because when we try to create request 2 it will do it with the same name as request 1.

Reply:
Ok, I understand. Then it will always be "_problematic_" to create several records at once (multi-create) because the sequence (or other field that is changed) will not be respected.

Reply:
Not really. Your code is not creating in batch. It's serializing creations. One example of batch would be:

```python
requests = self.env["stock.request"].create([{"product_id": 1}, {"product_id": 1}])
```

A problematic code in batch would be:

```python
vals = {"product_id": 1}
requests = self.env["stock.request"].create([vals, vals])
```

But as said, doing such technique with Odoo core you are exposed to the same problems, and it's very clear that it's you responsibility to don't misuse the call to ORM API methods that way. One proper code would be:

```python
vals = {"product_id": 1}
requests = self.env["stock.request"].create([vals.copy(), vals.copy()])
```

but it's very unfrequent that you need something like that. A typical example for requiring base values would be:

```python
base_vals = {"product_id": 1}
vals_list = []
for record in source:
    vals = base_vals.copy()
    vals["x"] = record.x
    vals_list.append(vals)
requests = self.env["stock.request"].create(vals_list)
```

and as said, you should do the copy, not the `create` override.

Reply:
> I don't see any problem here in the "upper stack". Can you point to any real example of a problem? This is causing a non-needed overhead IMO.

Don't know if it is still the case today, but with previous Odoo versions, when it retried a transaction,  it did so with the modified vals, which could lead to very subtle hard to debug issues.

Reply:
Do you mean that the whole create method is repeated? Do you have the piece of code that did/does that?

Reply:
@pedrobaeza it still does that in 16, it seems. See here, and the `retrying` function.

https://github.com/odoo/odoo/blob/8ecf71ff080d09461c72724f634c24a40e60e836/odoo/service/model.py#L50

So if you have things like postgres deadlocks, the whole function is retried with the same arguments. If you have modified arguments contents (like the vals dict for create and write), it is retried with the modified content.

I have had to diagnose that kind of heisenbug a couple of times and it's been very very difficult to track down.

To me, the Odoo json/xmlrpc interface should deserialize dicts as frozendicts... But in general, yes, I really recommend to avoid modifying function arguments objects.

Reply:
Thanks for the extra bits. Yes, I see the potential problem, but in this case, the method is re-entrant, as if you retry, the included conditions makes to not change the second/third time calling it.

The solution for me is not to pass them as frozendicts, but to make a copy of the arguments before calling the server method.

And as said, Odoo has the same flaws in its own code.

Did you talk about this problem with someone in Odoo?

Reply:
And complementing, this is only happen for direct `create` calls from the web client, not on normal business code calling `create` (although the business code have to take care about not reusing the same dictionary if it's making subsequent calls of create as stated before).